### PR TITLE
feat: support V2 OperationsClient in OperationResponse

### PR DESCRIPTION
**Pull request by ajupazhamayil**

### CODE REVIEW

**Review by bshaffer — CHANGES_REQUESTED** [review]
We are missing tests for this, and definitely want to add additional testing using the new V2 client in [OperationResponseTest](https://github.com/googleapis/gax-php/blob/main/tests/Tests/Unit/OperationResponseTest.php)

Additionally, we need to consider how this will work for `$this->additionalArgs` (e.g.  `Compute`), as these additional args need to be set on the `Request` object for V2, not to the RPC call. 

We probably need to modify the way the additional args are passed in from GAX. For example, in [Compute's `descriptor_config.php`](https://github.com/googleapis/google-cloud-php-compute/blob/db27fb2832731a827957c3adbaeb5983c1202a9b/src/V1/resources/disks_descriptor_config.php#L428-L431), we have the following:

```php
'SomeRPCMethod' => [
    'longRunning' => [
        'additionalArgumentMethods' => [
            'getProject',
            'getZone',
        ],
        // ...
    ]
]
```

This is then passed into the `OperationResponse` class as `$additionalArgs` by GAX:

https://github.com/googleapis/gax-php/blob/9b06449585573f76540cb55d8d0d041e232057ce/src/GapicClientTrait.php#L747 

But since we need to set these on the Request object, we could do something like to pass `$additionalArgs` in as a key-value pairs or setters, and then have those used in `OperationResponse`. 

*Example* (key-value pairs):

```
// in descriptor_config.php
        'additionalArgumentMethods' => [
            'project' => 'getProject',
            'zone' => 'getZone',
        ],

// in OperationResponse::operationsCall
new $requestClass(['name' => $this->getName()] + $this->additionalArgs);
```

*Example* (setters):

```
// in descriptor_config.php
        'additionalArgumentMethods' => [
            'setProject' => 'getProject',
            'setZone' => 'getZone',
        ],

// in OperationResponse::operationsCall
$request = $requestClass::build($this->getName());
foreach ($this->additionalArgs as $setter => $value) {
    $request->$setter($value);
}
```

**Comment by ajupazhamayil** [comment]
> ```
> setter
> ```


I see the request objects that we are referring in the [OperationsClient](https://github.com/googleapis/php-longrunning/blob/main/src/LongRunning/Client/OperationsClient.php) does not have any arguments other than the name. Example, see the [`DeleteOperationRequest`](https://github.com/googleapis/php-longrunning/blob/main/src/LongRunning/DeleteOperationRequest.php). I am unable to understand how these extra attributes are being set to those request methods!

File: src/OperationResponse.php [hunk 1]
```diff
@@ -386,7 +390,11 @@ public function cancel()
         if (is_null($this->cancelOperationMethod)) {
             throw new LogicException('The cancel operation is not supported by this API');
         }
-        $this->operationsCall($this->cancelOperationMethod, $this->getName(), $this->additionalArgs);
+        $this->operationsCall(
+            $this->cancelOperationMethod,
+            $this->buildOperationMethodArgument(CancelOperationRequest::class),
+            $this->additionalArgs
+        );
```
[the remark this answers is not in the document]
Comment (ajupazhamayil) [reply src/OperationResponse.php:397]: sounds great.

**Comment by ajupazhamayil** [comment]
Are they(`Compute` or similar) using a different client like [ZoneOperationsClient](https://github.com/googleapis/google-cloud-php-compute/blob/db27fb2832731a827957c3adbaeb5983c1202a9b/src/V1/Client/ZoneOperationsClient.php)?

File: src/OperationResponse.php [hunk 2]
```diff
@@ -483,4 +495,20 @@ private function hasProtoResponse()
...
     }
+
+    /**
+     * Create the argument for the operation.
+     *
+     * @param mixed $requestClass The class of the request to use for the operation.
+     *
+     * @return mixed The argument to use for the operation.
+     */
+    private function buildOperationMethodArgument(mixed $requestClass)
```
[the remark this answers is not in the document]
Comment (ajupazhamayil) [reply src/OperationResponse.php:506]: I will remove this method

**Comment by ajupazhamayil** [comment]
Please note tests are not yet added, working on it currently

File: tests/Tests/Unit/OperationResponseTest.php [hunk 3]
```diff
@@ -353,6 +358,26 @@ public function testPollingCastToInt()
...
+        $operationClient->deleteOperation(Argument::exact($request))
+            ->shouldBeCalledOnce();
+        $request = new CancelOperationRequest(['name' => 'test-123']);
+        $operationClient->cancelOperation(Argument::exact($request))
+            ->shouldBeCalledOnce();
+
+        $operationResponse = new OperationResponse('test-123', $operationClient->reveal());
+        $operationResponse->reload();
+        $operationResponse->delete();
+        $operationResponse->cancel();
```
Comment (bshaffer) [inline tests/Tests/Unit/OperationResponseTest.php:378]: nit: let's make these separate test cases

**Review by bshaffer — COMMENTED** [review]
This looks great! 

For testing, we could make all tests which do not mock specific operationClient methods accept a data provider for both V1 and V2 clients. I tested this out and it worked [see this diff](https://gist.github.com/bshaffer/31ed1a6eea5a907251eef9324214a470)

File: tests/Tests/Unit/OperationResponseTest.php [hunk 4]
```diff
@@ -353,6 +358,26 @@ public function testPollingCastToInt()
         $this->assertEquals($op->getSleeps(), [3, 4, 6]);
     }
 
+    public function testLROOperationsClient()
+    {
+        $operationClient = $this->prophesize(LROOperationsClient::class);
+        $request = new GetOperationRequest(['name' => 'test-123']);
+        $operationClient->getOperation(Argument::exact($request))
```
[the remark this answers is not in the document]
Comment (ajupazhamayil) [reply tests/Tests/Unit/OperationResponseTest.php:365]: Just now tested it and yes, `exact` is not required, removed the `exact`. thanks

File: tests/Tests/Unit/OperationResponseTest.php (hunk 3 above)
Comment (ajupazhamayil) [reply tests/Tests/Unit/OperationResponseTest.php:407]: Moved. thank you

File: src/OperationResponse.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (ajupazhamayil) [reply src/OperationResponse.php]: Removed

**Review by bshaffer — APPROVED** [review]

**Pull request merged**